allennlp/training/callbacks/callbacks.py

```diff
@@ -0,0 +1,512 @@
+            if self.tensorboard._histogram_interval is not None:
+                self.tensorboard.enable_activation_logging(trainer.model)
+
+        elif event == Events.BATCH_START and self.tensorboard.should_log_histograms_this_batch():
+            # Get the magnitude of parameter updates for logging
+            # We need a copy of current parameters to compute magnitude of updates,
+            # and copy them to CPU so large models won't go OOM on the GPU.
+            self.param_updates = {name: param.detach().cpu().clone()
+                                  for name, param in trainer.model.named_parameters()}
+
+        elif event == Events.BATCH_END:
+            # Log parameter values to tensorboard
+            if self.tensorboard.should_log_this_batch():
+                self.tensorboard.log_parameter_and_gradient_statistics(trainer.model, trainer.batch_grad_norm)
+                self.tensorboard.log_learning_rates(trainer.model, trainer.optimizer)
+
+                self.tensorboard.add_train_scalar("loss/loss_train", trainer.train_metrics["loss"])
+                self.tensorboard.log_metrics({"epoch_metrics/" + k: v for k, v in trainer.train_metrics.items()})
+
+
+            if self.log_batch_size_period:
+                cur_batch = sum([training_util.get_batch_size(batch) for batch in trainer.batch_group])
+                self.cumulative_batch_size += cur_batch
+                if (trainer.batches_this_epoch - 1) % self.log_batch_size_period == 0:
+                    average = self.cumulative_batch_size / trainer.batches_this_epoch
+                    logger.info(f"current batch size: {cur_batch} mean batch size: {average}")
+                    self.tensorboard.add_train_scalar("current_batch_size", cur_batch)
+                    self.tensorboard.add_train_scalar("mean_batch_size", average)
+
+            if self.tensorboard.should_log_histograms_this_batch():
+                for name, param in trainer.model.named_parameters():
+                    self.param_updates[name].sub_(param.detach().cpu())
+                    update_norm = torch.norm(self.param_updates[name].view(-1, ))
+                    param_norm = torch.norm(param.view(-1, )).cpu()
+                    self.tensorboard.add_train_scalar("gradient_update/" + name,
+                                                      update_norm / (param_norm + 1e-7))
+                self.param_updates.clear()
+                self.tensorboard.log_histograms(trainer.model, self.histogram_parameters)
+
+        elif event == Events.EPOCH_END:
+            self.tensorboard.log_metrics(trainer.train_metrics,
+                                         val_metrics=trainer.val_metrics,
+                                         log_to_console=True,
+                                         epoch=trainer.epoch_number + 1)
+
+    @classmethod
+    def from_params(cls, serialization_dir: str, params: Params) -> 'LogToTensorboard':  # type: ignore
```

Comment:
Why do you need this?  Oh, because `serialization_dir` isn't a `TensorboardWriter`?  It'd be nice to remove the need for this custom `from_params` somehow.